File: Leetcode/图论/树的直径.java
class Solution {
int res = 0;
    public int treeDiameter(int[][] edges) {
        List<Integer>[] map = new ArrayList[edges.length+1];
        
        for(int i=0 ; i<map.length ; i++){
            map[i] = new ArrayList<>();
        }
        for(int[] edge : edges){
            map[edge[0]].add(edge[1]);
            map[edge[1]].add(edge[0]);
        }
        
        dfs(map,0,new boolean[edges.length+1]);
        return res;
    }
    
    public int dfs(List<Integer>[] map,int index,boolean[] visited){
        visited[index] = true;
        List<Integer> list = map[index];
        int max1 = 0;   // 第一大
        int max2 = 0;   // 第二大
        for(int next : list){
            if(!visited[next]){               
                int num = dfs(map,next,visited);
                if(num>max1){
                    max2 = max1;
                    max1 = num;
                } else if(num>max2){
                    max2 = num;
                }
            }
        }
        res = Math.max(res,max1+max2);
        
        return Math.max(max1,max2)+1;
        
    }
}

File: Leetcode/图论/树的直径.cpp
//两次dfs/bfs
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int treeDiameter(vector<vector<int>>& edges) {
        unordered_map<int,vector<int>> m;
        int maxnode = 0;
        for(auto edge:edges){
            m[edge[0]].push_back(edge[1]);
            m[edge[1]].push_back(edge[0]);
            maxnode = max(maxnode,edge[0]);
            maxnode = max(maxnode,edge[1]);
        }
        vector<int> visited(maxnode+1,false);

        queue<pair<int,int>> q;
        q.push({0,0});
        int res=0 ;
        int n1 = 0;
        while(!q.empty()){
            auto t = q.front();q.pop();
            if(visited[t.first]) continue;
            visited[t.first] = true;
            if(t.second > res){
                res = t.second;
                n1 = t.first;
            }
            for(int x:m[t.first]){
                if(!visited[x])
                    q.push({x,t.second + 1});
            }
        }

        // cout<<n1<<endl;

        while(!q.empty()) q.pop();

        visited=vector<int>(maxnode+1,false);
        q.push({n1,0});
        while(!q.empty()){
            auto t = q.front();q.pop();
            if(visited[t.first]) continue;
            visited[t.first] = true;
            if(t.second>res){
                res = t.second;
                n1 = t.first;
            }
            for(int x:m[t.first]){
                if(!visited[x])
                    q.push({x,t.second+1});
            }
        }
        return res;
    }  
};

//树形dp